From 879eb9027611f215441ae51151b387d55708750c Mon Sep 17 00:00:00 2001
From: Jonas Jelonek <jelonek.jonas@gmail.com>
Date: Tue, 10 Feb 2026 08:27:34 +0000
Subject: [PATCH] net: phy: be more graceful when reading PHY IDs

Some Broadcom PHYs have been seen which behave quite nasty when
accessing different MMDs. While the corresponding bits in various
"Devices in package" are set, the MMDs later don't react to reading the
"Device present" bits or even the PHY ID. Still, those MMDs are there,
otherwise functional and required by a driver. However, the current PHY
ID reading procedure is rather harsh when this situation is encountered,
leading to a hard fail.

Adjust that to just continue with other MMDs instead of cancelling the
whole probing. This way, those MMDs do not show a valid PHY ID but those
MMDs are still considered as usable.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

--- a/drivers/net/phy/phy_device.c
+++ b/drivers/net/phy/phy_device.c
@@ -876,7 +876,8 @@ static int get_phy_c45_ids(struct mii_bu
 {
 	const int num_ids = ARRAY_SIZE(c45_ids->device_ids);
 	u32 devs_in_pkg = 0;
-	int i, ret, phy_reg;
+	int i, ret, phy_reg, phy_reg2;
+	u32 devs_in_pkg = 0;
 
 	/* Find first non-zero Devices In package. Device zero is reserved
 	 * for 802.3 c45 complied PHYs, so don't probe it at first.
@@ -931,22 +932,23 @@ static int get_phy_c45_ids(struct mii_bu
 			 * registers.
 			 */
 			ret = phy_c45_probe_present(bus, addr, i);
-			if (ret < 0)
-				return ret;
-
-			if (!ret)
+			if (ret <= 0)
 				continue;
 		}
 
+		/* Do not fail if we can't read the ID. Some PHYs have
+		 * MMDs here which do not react to the device present
+		 * bits nor the PHY ID but are still "there" and needed
+		 * by the driver.
+		 */
 		phy_reg = mdiobus_c45_read(bus, addr, i, MII_PHYSID1);
-		if (phy_reg < 0)
-			return -EIO;
-		c45_ids->device_ids[i] = phy_reg << 16;
+		phy_reg2 = mdiobus_c45_read(bus, addr, i, MII_PHYSID2);
+		if (phy_reg < 0 || phy_reg2 < 0) {
+			c45_ids->device_ids[i] = 0xffffffff;
+			continue;
+		}
 
-		phy_reg = mdiobus_c45_read(bus, addr, i, MII_PHYSID2);
-		if (phy_reg < 0)
-			return -EIO;
-		c45_ids->device_ids[i] |= phy_reg;
+		c45_ids->device_ids[i] = phy_reg << 16 | phy_reg2;
 	}
 
 	c45_ids->devices_in_package = devs_in_pkg;
